h/static/scripts/annotator/range-util.js

```diff
@@ -0,0 +1,146 @@
+    return selection.focusOffset < selection.anchorOffset;
+  }
+
+  var range = selection.getRangeAt(0);
+  return range.startContainer === selection.focusNode;
+}
+
+/**
+ * Returns true if `node` is between the `startContainer` and `endContainer` of
+ * the given `range`, inclusive.
+ *
+ * @param {Range} range
+ * @param {Node} node
+ */
+function isNodeInRange(range, node) {
+  if (node === range.startContainer || node === range.endContainer) {
+    return true;
+  }
+
+  /* jshint -W016 */
+  var isAfterStart = range.startContainer.compareDocumentPosition(node) &
+   Node.DOCUMENT_POSITION_FOLLOWING;
+  var isBeforeEnd = range.endContainer.compareDocumentPosition(node) &
+   Node.DOCUMENT_POSITION_PRECEDING;
+
+  return isAfterStart && isBeforeEnd;
+}
+
+/**
+ * Iterate over all Node(s) in `range` in document order and invoke `callback`
+ * for each of them.
+ *
+ * @param {Range} range
+ * @param {Function} callback
+ */
+function forEachNodeInRange(range, callback) {
+  var root = range.commonAncestorContainer;
+
+  // The `whatToShow`, `filter` and `expandEntityReferences` arguments are
+  // mandatory in IE although optional according to the spec.
+  var nodeIter = root.ownerDocument.createNodeIterator(root,
+    NodeFilter.SHOW_ALL, null /* filter */, false /* expandEntityReferences */);
+
+  /* jshint -W084 */
+  var currentNode;
+  while (currentNode = nodeIter.nextNode()) {
+    if (isNodeInRange(range, currentNode)) {
+      callback(currentNode);
+    }
+  }
+}
+
+/**
+ * Returns the bounding rectangles of non-whitespace text nodes in `range`.
+ *
+ * @param {Range} range
+ * @return {Array<Rect>} Array of bounding rects in document coordinates.
+ */
+function getTextBoundingBoxes(range) {
+  var whitespaceOnly = /^\s*$/;
+  var textNodes = [];
+  forEachNodeInRange(range, function (node) {
+    if (node.nodeType === Node.TEXT_NODE &&
+        !node.textContent.match(whitespaceOnly)) {
+      textNodes.push(node);
+    }
+  });
+
+  var rects = [];
+  textNodes.forEach(function (node) {
+    var nodeRange = node.ownerDocument.createRange();
+    nodeRange.selectNodeContents(node);
+    if (node === range.startContainer) {
+      nodeRange.setStart(node, range.startOffset);
+    }
+    if (node === range.endContainer) {
+      nodeRange.setEnd(node, range.endOffset);
+    }
+    if (nodeRange.collapsed) {
+      // If the range ends at the start of this text node or starts at the end
+      // of this node then do not include it.
+      return;
+    }
+
+    // Measure the range and translate from viewport to document coordinates
+    var viewportRects = Array.from(nodeRange.getClientRects());
+    nodeRange.detach();
+    rects = rects.concat(viewportRects.map(function (rect) {
+      return mapViewportRectToDocument(node.ownerDocument.defaultView, rect);
+    }));
+  });
+  return rects;
+}
+
+/**
+ * Returns the rectangle, in document coordinates, for the line of text
+ * containing the focus point of a Selection.
```

Comment:
I'm wondering about a better name for this. I'm not entirely sure what focus rect / point means. I think what this function does is return the bounding box of the "last" word in the selection, where the definition of "last" depends on (our guess at) which direction the selection is made in? In that case `lastBoundingBox(selection)` might be a better name?

Also the function name says focus rect but the docstring says focus point, seems like the docstring is wrong.